horovod/torch/tensor_util.h

```diff
@@ -35,171 +35,211 @@ namespace torch {
+  static void DivideTensorInPlace(T* tensor, int value);
 
 #if HAVE_CUDA
-  template <class T, class TC> static void CopyCPUToCuda(T* cpu, TC* cuda);
-  template <class TC, class T> static void AsyncCopyCudaToCPU(TC* cuda, T* cpu);
+  template <MPIDataType DT, class T, class TC>
+  static void CopyCPUToCuda(T* cpu, TC* cuda);
+  template <MPIDataType DT, class TC, class T>
+  static void AsyncCopyCudaToCPU(TC* cuda, T* cpu);
 #endif
 };
 
-#define TENSOR_UTIL_DEFINE_TYPE_H(THTensor)                                    \
-  template <> const MPIDataType TensorUtil::GetDType<THTensor>();              \
+#define TENSOR_UTIL_DEFINE_TYPE_H(HorovodType, DeviceType, THTensor)           \
+  template <>                                                                  \
+  const TensorShape TensorUtil::GetShape<HorovodType, DeviceType, THTensor>(   \
+      THTensor * tensor);                                                      \
+  template <>                                                                  \
+  const void* TensorUtil::GetData<HorovodType, DeviceType, THTensor>(          \
+      THTensor * tensor);                                                      \
   template <>                                                                  \
-  const TensorShape TensorUtil::GetShape<THTensor>(THTensor * tensor);         \
-  template <> const void* TensorUtil::GetData<THTensor>(THTensor * tensor);    \
-  template <> int64_t TensorUtil::GetSize<THTensor>(THTensor * tensor);        \
-  template <> int TensorUtil::GetDevice<THTensor>(THTensor * tensor);          \
+  int64_t TensorUtil::GetSize<HorovodType, DeviceType, THTensor>(THTensor *    \
+                                                                 tensor);      \
+  template <>                                                                  \
+  int TensorUtil::GetDevice<HorovodType, DeviceType, THTensor>(THTensor *      \
+                                                               tensor);        \
                                                                                \
-  template <> THTensor* TensorUtil::New<THTensor>(int device);                 \
-  template <> void TensorUtil::Free<THTensor>(THTensor * tensor);              \
   template <>                                                                  \
-  void TensorUtil::ResizeNd<THTensor>(THTensor * tensor, int nDimension,       \
-                                      int64_t* size, int64_t* stride);         \
+  THTensor* TensorUtil::New<HorovodType, DeviceType, THTensor>(int device);    \
+  template <>                                                                  \
+  void TensorUtil::Free<HorovodType, DeviceType, THTensor>(THTensor * tensor); \
+  template <>                                                                  \
+  void TensorUtil::ResizeNd<HorovodType, DeviceType, THTensor>(                \
+      THTensor * tensor, int nDimension, int64_t* size, int64_t* stride);      \
```

Comment:
Nit: Why the inconsistent style with `THTensor *` and `int64_t*`?

Reply:
No idea - `clang-format` does it :-)

Reply:
May as well roll with it then.